pulse2percept/models/granley2021.py

```diff
@@ -0,0 +1,559 @@
+        Use probabilistic sigmoid thresholding, default: False
+    **params: dict, optional
+        Additional params for AxonMapModel. 
+
+        Options for AxonMapSpatial:
+        ---------------------------
+        axlambda: double, optional
+            Exponential decay constant along the axon(microns).
+        rho: double, optional
+            Exponential decay constant away from the axon(microns).
+        eye: {'RE', LE'}, optional
+            Eye for which to generate the axon map.
+        xrange : (x_min, x_max), optional
+            A tuple indicating the range of x values to simulate (in degrees of
+            visual angle). In a right eye, negative x values correspond to the
+            temporal retina, and positive x values to the nasal retina. In a left
+            eye, the opposite is true.
+        yrange : tuple, (y_min, y_max)
+            A tuple indicating the range of y values to simulate (in degrees of
+            visual angle). Negative y values correspond to the superior retina,
+            and positive y values to the inferior retina.
+        xystep : int, double, tuple
+            Step size for the range of (x,y) values to simulate (in degrees of
+            visual angle). For example, to create a grid with x values [0, 0.5, 1]
+            use ``x_range=(0, 1)`` and ``xystep=0.5``.
+        grid_type : {'rectangular', 'hexagonal'}
+            Whether to simulate points on a rectangular or hexagonal grid
+        loc_od, loc_od: (x,y), optional
+            Location of the optic disc in degrees of visual angle. Note that the
+            optic disc in a left eye will be corrected to have a negative x
+            coordinate.
+        n_axons: int, optional
+            Number of axons to generate.
+        axons_range: (min, max), optional
+            The range of angles(in degrees) at which axons exit the optic disc.
+            This corresponds to the range of $\\phi_0$ values used in
+            [Jansonius2009]_.
+        n_ax_segments: int, optional
+            Number of segments an axon is made of.
+        ax_segments_range: (min, max), optional
+            Lower and upper bounds for the radial position values(polar coords)
+            for each axon.
+        min_ax_sensitivity: float, optional
+            Axon segments whose contribution to brightness is smaller than this
+            value will be pruned to improve computational efficiency. Set to a
+            value between 0 and 1.
+        axon_pickle: str, optional
+            File name in which to store precomputed axon maps.
+        ignore_pickle: bool, optional
+            A flag whether to ignore the pickle file in future calls to
+            ``model.build()``.
+    """
+
+    def __init__(self, **params):
+        super(BiphasicAxonMapSpatial, self).__init__(**params)
+
+    def __getattr__(self, attr):
+        # Called when normal get attribute fails
+        if sys._getframe(2).f_code.co_name == '__init__' or  \
```

Comment:
Should this be added to the `Frozen` class, or does it just apply here because you're nesting models/classes?

Reply:
It applies here because I overrided __setattr__, effectively adding one stack to the frame. Thus init appears in position 3 instead of position 2. I think it should be okay for the frozen class though, because if init is anywhere in the call stack frame, then this was still part of the constructor.

Reply:
Just to clarify, i modified the code to print out the top 5 frames when `__setattr__` is called. 
AxonMapModel:
`['set_attr', '__init__', '__init__', '__init__', '__init__']`
BiphasicAxonMapModel:
`['set_attr', '__setattr__', '__init__', '__init__', '__init__']`
Because of the way the frozen class and nested model classes work, most subclasses never call their own __setattr__, they just use the frozen one from the parent class instead. Since I need to distribute calls to `__setattr__` to the relevant effects models, BiphasicAxonMapSpatial implements its own `__setattr__`, which adds the call frame to the stack

Reply:
Gotcha - that makes sense